client/asset/eth/eth.go

```diff
@@ -4596,177 +4516,617 @@ func (w *baseWallet) emitTransactionNote(tx *asset.WalletTransaction, new bool)
+		}
+		newPendingTx := w.extendedTx(replacementTx, asset.Unknown, 0)
+		pendingTx.NonceReplacement = newPendingTx.ID
+		var oldTo, newTo common.Address
+		if oldAddr := oldTx.To(); oldAddr != nil {
+			oldTo = *oldAddr
+		}
+		if newAddr := replacementTx.To(); newAddr != nil {
+			newTo = *newAddr
+		}
+		if bytes.Equal(oldTx.Data(), replacementTx.Data()) && oldTo == newTo {
+			pendingTx.FeeReplacement = true
+		}
+		w.tryStoreDBTx(pendingTx)
+		w.tryStoreDBTx(newPendingTx)
+		w.pendingTxs[idx] = newPendingTx
+		return nil
+	})
 }
 
-// checkPendingTxs checks the confirmation status of all pending transactions.
-func (w *baseWallet) checkPendingTxs() {
-	pendingTxsCopy := make(map[uint64]*extendedWalletTx, len(w.pendingTxs))
-	w.pendingTxsMtx.Lock()
-	for nonce, tx := range w.pendingTxs {
-		pendingTxsCopy[nonce] = tx
+// userActionRecoverNonces, if recover is true, examines our confirmed and
+// pending nonces and our pendingTx set and sends zero-value txs to ourselves
+// to fill any gaps or replace any rogue transactions. This should never happen
+// if we're only running one instance of this wallet.
+func (w *assetWallet) userActionRecoverNonces(actionB []byte) error {
+	var action struct {
+		Recover *bool `json:"recover"`
+	}
+	if err := json.Unmarshal(actionB, &action); err != nil {
+		return fmt.Errorf("error unmarshaling user action: %v", err)
+	}
+	if action.Recover == nil {
+		return errors.New("no recover value specified")
+	}
+	if !*action.Recover {
+		// Don't reset recoveryRequestSent. They won't see this message again until
+		// they reboot.
+		return nil
+	}
+	maxFeeRate, tipRate, err := w.recommendedMaxFeeRate(w.ctx)
+	if err != nil {
+		return fmt.Errorf("error getting max fee rate for nonce resolution: %v", err)
+	}
+	w.nonceMtx.Lock()
+	defer w.nonceMtx.Unlock()
+	missingNonces := findMissingNonces(w.confirmedNonceAt, w.pendingNonceAt, w.pendingTxs)
+	if len(missingNonces) == 0 {
+		return nil
+	}
+	for i, n := range missingNonces {
+		nonce := new(big.Int).SetUint64(n)
+		txOpts, err := w.node.txOpts(w.ctx, 0, defaultSendGasLimit, maxFeeRate, tipRate, nonce)
+		if err != nil {
+			return fmt.Errorf("error getting tx opts for nonce resolution: %v", err)
+		}
+		var skip bool
+		tx, err := w.node.sendTransaction(w.ctx, txOpts, w.addr, nil, func(err error) (discard, propagate, fail bool) {
+			if errorFilter(err, "replacement transaction underpriced") {
+				skip = true
+				return true, false, false
+			}
+			return false, false, true
+		})
+		if err != nil {
+			return fmt.Errorf("error sending tx %d for nonce resolution: %v", nonce, err)
+		}
+		if skip {
+			w.log.Warnf("skipping storing underpriced replacement tx for nonce %d", nonce)
+		} else {
+			w.pendingTxs = append(w.pendingTxs, w.extendAndStoreTx(tx, asset.SelfSend, 0, nil))
+			sort.Slice(w.pendingTxs, func(i, j int) bool {
+				return w.pendingTxs[i].Nonce.Cmp(w.pendingTxs[j].Nonce) < 0
+			})
```

Comment:
Emit a transaction note here? Currently the recovery transactions don't show up until you refresh.